### DISCUSSION TITLE
Passing the dependency to the child State

### DISCUSSION THREAD

**Discussion by zvona031** [pointfreeco/swift-dependencies#112] [Q&A]
I am using Dependencies library in combination with The Composable Architecture and I ran into an issue that I need help with.
I have a ParentFeature, whose State has 2 ChildFeature.State-s. ChildFeature has a @Dependency. Now I am not sure how to properly change the child features dependencyes when creating ParentFeature.

```
struct ChildFeature: Reducer {
    struct State {}
    
    enum Action {}
    
    @Dependency(\.calendar) var calendar
    
    var body: some ReducerOf<Self> {
        Reduce { state, action in
            return .none
        }
    }
}
```

```
struct ParentFeature: Reducer {
    struct State {
        var A: ChildFeature.State
        var B: ChildFeature.State
    }
    
    enum Action {
        case aActins(ChildFeature.Action)
        case bActions(ChildFeature.Action)
    }
    
    var body: some ReducerOf<Self> {
        Scope(state: \.A, action: /Action.aActins) {
            ChildFeature()
//                .dependency(\.calendar, MyCustomCalendar) ????
        }
        
        Scope(state: \.A, action: /Action.aActins) {
            ChildFeature()
//                .dependency(\.calendar, MyAnotherCustomCalendar) ????
        }
        Reduce { state, action in
            return .none
        }
    }
}
```

**Comment by mbrandonw** [comment]
Hi @zvona031, it's not clear what you are trying to accomplish. The code you have written:

```swift
ChildFeature()
  .dependency(\.calendar, …)
```

…is one way you can override the dependency for a child feature. What problems are you having with this?

**Reply by zvona031** [reply]
@mbrandonw The problem that I would like to solve is the following. I want to test the ParentFeature, in a way that one ChildFeature has one dependency, and the other ChildFeature has a completely different dependency. Is it possible to achieve something like that?
But then the question emerges, is it even necessary to create a test case for that situation, or the Child dependencies should not be touched when testing the Parent?

**Reply by mbrandonw** [reply]
I'm still not entirely sure of the situation you are dealing with, but it sounds like you may want two separate dependencies. You may need _two_ calendars registered in the dependency system, `\.calendarA` and `\.calendarB`, so that each feature can use one and you can control them independently for tests. 

> But then the question emerges, is it even necessary to create a test case for that situation, or the Child dependencies should not be touched when testing the Parent?

If the dependency is not touched during a test then there is no need to override it in tests.

**Reply by zvona031** [reply]
I would try to explain the real situation that I am dealing with. So the ChildFeature represents the InputView, and childs dependency is a method that validates the text from InputView. So in that way, I can have different validations for the InputViews. The ParentFeature is actually a screen where I have multiple InputViews (some kind of a form). In the body of a ParentFeature, I override dependencies for each InputView with a different custom value, because I have a different type of validations for each InputView. 
For the live value of that validation method, I have some kind of EmptyValidator (Input View is always valid, no matter the text inside of it). So in my screen, the first InputView has the live value dependency, the second one has custom value (Input view text has to have at least one character), and the third one has to pass the email regex.
And it works fine when I build it. 
But when I want to recreate that situation while testing the ParentFeature, I am not sure how to do that. Because I am not sure how to change the dependency for each InputView inside of my ParentView. 
When I try to test it, all the InputViews have the same dependencies as in the production build, even if I make the TestStore using the withDependencies helper.
